### ISSUE TITLE
Enforce HTTPS for dist URL

### ISSUE THREAD

**Issue by mbrodala** [composer/composer#4959]
My `composer.lock` currently contains various dist URLs from composer.typo3.org which use HTTP instead of HTTPS which causes issues with the latest Composer version. The dist URLs on composer.typo3.org are all HTTPS but where HTTP previously.

Instead of setting `secure-http` to `false` I'd prefer if Composer would simply switch from HTTP to HTTPS. How can I enforce this?

Even if I fix the URLs manually to HTTPS, Composer reverts my change on `composer update --lock`.

**Comment by Seldaek** [comment]
There really isn't a great way for us to do this automatically as some domains might not have https. What you can do to fix it though is update the URLs in vendor/composer/installed.json and then `composer update --lock` should do the right thing.

**Comment by curry684** [comment]
@Seldaek isn't this part of a bigger issue that we have to detect http=>https redirects for several situations?

In this case properly detecting it could allow us to automatically update URLs, in preparation of the http ones when a remote repo updates their URLs. The second use case is the issue which appears to keep popping up that in case of an upgrade redirect we use the HTTP_PROXY values to resolve HTTPS URLs even if HTTPS_PROXY is defined.

**Comment by mbrodala** [comment]
Thanks, @Seldaek, that fixed the issue for me. An automatic update as suggested by @curry684 would be nice indeed.

**Comment by Seldaek** [comment]
@curry684 I don't think this is related no, there is no redirection here, just an outdated installation and a repo that used http URLs before (which just became forbidden by #4907)

**Comment by curry684** [comment]
Sorry didn't mean to imply it's related, but just in general a good idea to handle redirects on our end instead of automagically, as it would allow us to correct URLs in config as well if a 301 (Permanent Redirect) is issued.

Doesn't matter in this case I see as http://composer.typo3.org/packages.json is still available on http, so no way for us to detect it other than guesswork.

**Comment by mbrodala** [comment]
@curry684 Thanks for the heads up, we'll put a redirect to HTTPS in place.

**Comment by StormTide** [comment]
FYI, a redirect follow from HTTP to HTTPS isnt safe (MITM attacker controls the redirect, and can pick a different hostname that supports HTTPS but isnt the intended). In this case the installed.json really needs to be updated with the correct https urls.

**Comment by curry684** [comment]
Yeah sorry I did intend to add updates should only be implicit if same-domain. Could still have some loopholes but nothing nastier than a MITM could do if continuing to use HTTP, which we already forbid now.

So as a whole there's nothing really left to do in this issue as the secure-http default is already in prod.

**Comment by StormTide** [comment]
For clarity, even on the same domain it would be bad. Consider a http redirect involving github.com, could redirect any package to any other package, including one of the attackers control.  Could check the entire URL with http vs https or something, but even that im sure has a regression somewhere where the scheme is a different site (think ip name based virtual hosting on http where the https is controlled by another client)... 

Bottom line, for secure operation, the package url has to be https, and that's now enforced in prod -- and thats a good thing, even if its a pain to update the packaging json files.
